File: go/pkg/antflylite/status.go
// Copyright 2026 Antfly, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//go:build cgo

package antflylite

import "encoding/json"

// StorageStatus describes the physical Lite database file backing a handle.
type StorageStatus struct {
	Format             string  `json:"format"`
	Engine             string  `json:"engine"`
	PrimaryLayout      string  `json:"primary_layout"`
	ReplayLayout       string  `json:"replay_layout"`
	IndexLayout        string  `json:"index_layout"`
	IndexNamespace     *string `json:"index_namespace,omitempty"`
	FormatVersion      *uint32 `json:"format_version,omitempty"`
	PageSize           *uint32 `json:"page_size,omitempty"`
	ActiveCheckpoint   *uint8  `json:"active_checkpoint,omitempty"`
	CheckpointSequence *uint64 `json:"checkpoint_sequence,omitempty"`
	PageCount          *uint64 `json:"page_count,omitempty"`
}

// InferenceStatus reports the configured inference execution mode for a Lite
// handle. A fresh core Lite database normally reports Configured=false and
// NoInferenceConfiguredOK=true.
//
// HostBudgetMB, BackendBudgetMB, CombinedBudgetMB, KVBudgetMB,
// ScratchBudgetMB, and ProcessMemoryBudgetMB initially echo the OpenOptions
// resource-budget overrides requested for the local embedded runtime (0
// meaning the embedded node's own host-clamped default, not automatic/
// unbounded). Once a local runtime actually starts, all six are overwritten
// with the resolved effective values the node installed -- either the
// explicit override, or the host-clamped default from antfly's
// inference_provider.zig (default_host_budget_mb and friends) -- so none of
// the five generation-budget fields read 0 for a started local runtime.
// ProcessMemoryLimitBytes and ProcessMemoryLimitSource report the process
// memory envelope the embedded node resolved: either the explicit
// ProcessMemoryBudgetMB override, or the same host/cgroup-detected policy
// `antfly inference run` and `antfly standalone` report at startup. All of
// these stay at their zero-value/"automatic" defaults until a local runtime
// is actually started.
type InferenceStatus struct {
	Mode                     string   `json:"mode"`
	AvailableModes           []string `json:"available_modes"`
	Configured               bool     `json:"configured"`
	RemoteProviderConfigured bool     `json:"remote_provider_configured"`
	LocalRuntimeConfigured   bool     `json:"local_runtime_configured"`
	LocalRuntimeAvailable    bool     `json:"local_runtime_available"`
	CallerSuppliedArtifacts  bool     `json:"caller_supplied_artifacts"`
	NoInferenceConfiguredOK  bool     `json:"no_inference_configured_ok"`
	HostBudgetMB             uint32   `json:"host_budget_mb"`
	BackendBudgetMB          uint32   `json:"backend_budget_mb"`
	CombinedBudgetMB         uint32   `json:"combined_budget_mb"`
	KVBudgetMB               uint32   `json:"kv_budget_mb"`
	ScratchBudgetMB          uint32   `json:"scratch_budget_mb"`
	ProcessMemoryBudgetMB    uint32   `json:"process_memory_budget_mb"`
	ProcessMemoryLimitBytes  uint64   `json:"process_memory_limit_bytes"`
	ProcessMemoryLimitSource string   `json:"process_memory_limit_source"`
}

// Capabilities describes the Lite feature contract advertised by a handle.
type Capabilities struct {
	FreestandingBuild                  bool     `json:"freestanding_build"`
	HostedProfile                      bool     `json:"hosted_profile"`
	ManualMaintenance                  bool     `json:"manual_maintenance"`
	BackgroundEnrichmentRuntime        bool     `json:"background_enrichment_runtime"`
	TTLCleanupRuntime                  bool     `json:"ttl_cleanup_runtime"`
	TransactionRecoveryRuntime         bool     `json:"transaction_recovery_runtime"`
	LocalTemplateRendering             bool     `json:"local_template_rendering"`
	RemoteTemplateRendering            bool     `json:"remote_template_rendering"`
	RemoteTemplateHostCallbacks        bool     `json:"remote_template_host_callbacks"`
	InferenceMode                      string   `json:"inference_mode"`
	SupportedInferenceModes            []string `json:"supported_inference_modes"`
	AvailableInferenceModes            []string `json:"available_inference_modes"`
	InferenceRequired                  bool     `json:"inference_required"`
	NoInferenceConfiguredOK            bool     `json:"no_inference_configured_ok"`
	CallerSuppliedArtifacts            bool     `json:"caller_supplied_artifacts"`
	CallerSuppliedEmbeddings           bool     `json:"caller_supplied_embeddings"`
	RemoteInferenceProviders           bool     `json:"remote_inference_providers"`
	LocalInferenceRuntime              bool     `json:"local_inference_runtime"`
	GeneratedEnrichmentPlanning        bool     `json:"generated_enrichment_planning"`
	TextSearch                         bool     `json:"text_search"`
	DenseVectorSearch                  bool     `json:"dense_vector_search"`
	SparseVectorSearch                 bool     `json:"sparse_vector_search"`
	HybridSearch                       bool     `json:"hybrid_search"`
	GraphSearch                        bool     `json:"graph_search"`
	DistributedShardOwnership          bool     `json:"distributed_shard_ownership"`
	RaftReplication                    bool     `json:"raft_replication"`
	ClusterPlacement                   bool     `json:"cluster_placement"`
	CrossNodeJoins                     bool     `json:"cross_node_joins"`
	RemoteShardFanout                  bool     `json:"remote_shard_fanout"`
	DistributedTransactionCoordination bool     `json:"distributed_transaction_coordination"`
	ClusterHeartbeatStatusAggregation  bool     `json:"cluster_heartbeat_status_aggregation"`
	ServerSideAutoscaling              bool     `json:"server_side_autoscaling"`
	KubernetesOperator                 bool     `json:"kubernetes_operator"`
	ObjectStoragePrimary               bool     `json:"object_storage_primary"`
}

// PendingWorkStatus describes the stable Lite readiness fields for derived
// work. The nested maintenance telemetry is intentionally retained as raw JSON
// because it is operational detail rather than control-plane contract.
type PendingWorkStatus struct {
	DerivedTargetSequence               uint64          `json:"derived_target_sequence"`
	HasAsyncIndexes                     bool            `json:"has_async_indexes"`
	PortableImportPublicationInProgress bool            `json:"portable_import_publication_in_progress"`
	PortableImportRecoveryRequired      bool            `json:"portable_import_recovery_required"`
	PortableRuntimeActivationPending    bool            `json:"portable_runtime_activation_pending"`
	Enrichment                          json.RawMessage `json:"enrichment"`
	Resolution                          json.RawMessage `json:"resolution"`
	Promotion                           json.RawMessage `json:"promotion"`
	TextMerge                           json.RawMessage `json:"text_merge"`
}

// Status is the typed form of StatusJSON. Stats is retained as raw JSON because
// its internal shape is broader than the stable Lite control fields bindings
// need for feature branching.
type Status struct {
	Storage      StorageStatus     `json:"storage"`
	Stats        json.RawMessage   `json:"stats"`
	PendingWork  PendingWorkStatus `json:"pending_work"`
	Inference    InferenceStatus   `json:"inference"`
	Capabilities Capabilities      `json:"capabilities"`
}

// ReplayGeneratedEnrichmentsResult reports how many generated enrichment
// references were recreated from stored documents.
type ReplayGeneratedEnrichmentsResult struct {
	Replayed uint64 `json:"replayed"`
}

// Status returns the typed Lite status document for the database.
func (db *DB) Status() (*Status, error) {
	body, err := db.StatusJSON()
	if err != nil {
		return nil, err
	}
	var status Status
	if err := json.Unmarshal(body, &status); err != nil {
		return nil, err
	}
	return &status, nil
}

// PendingWorkStats returns the typed Lite pending-work readiness document.
func (db *DB) PendingWorkStats() (*PendingWorkStatus, error) {
	body, err := db.PendingWorkStatsJSON()
	if err != nil {
		return nil, err
	}
	var pending PendingWorkStatus
	if err := json.Unmarshal(body, &pending); err != nil {
		return nil, err
	}
	return &pending, nil
}

// RunUntilIdleStatus drains pending enrichment and index work and returns the
// typed post-drain readiness document.
func (db *DB) RunUntilIdleStatus() (*PendingWorkStatus, error) {
	body, err := db.RunUntilIdleJSON()
	if err != nil {
		return nil, err
	}
	var pending PendingWorkStatus
	if err := json.Unmarshal(body, &pending); err != nil {
		return nil, err
	}
	return &pending, nil
}

// ReplayGeneratedEnrichments recreates generated enrichment work from stored
// documents and returns the replay count.
func (db *DB) ReplayGeneratedEnrichments() (*ReplayGeneratedEnrichmentsResult, error) {
	body, err := db.ReplayGeneratedEnrichmentsJSON()
	if err != nil {
		return nil, err
	}
	var result ReplayGeneratedEnrichmentsResult
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

// Capabilities returns the typed Lite capability document for the database.
func (db *DB) Capabilities() (*Capabilities, error) {
	body, err := db.CapabilitiesJSON()
	if err != nil {
		return nil, err
	}
	var caps Capabilities
	if err := json.Unmarshal(body, &caps); err != nil {
		return nil, err
	}
	return &caps, nil
}

File: rs/crates/lite/src/status.rs
//! Typed convenience wrappers around the raw `*_json` status/capabilities
//! calls. Only compiled with the (default-on) `serde` feature. Mirrors
//! `go/pkg/lite/status.go`.

use std::fmt;

use serde::Deserialize;

use crate::db::Database;
use crate::error::Error;

/// Either an FFI error or a JSON decode failure from a typed convenience
/// method. The raw `*_json` methods on [`Database`] never produce this;
/// they return [`crate::Result`].
#[derive(Debug)]
pub enum TypedError {
    Ffi(Error),
    Json(serde_json::Error),
}

impl From<Error> for TypedError {
    fn from(err: Error) -> Self {
        TypedError::Ffi(err)
    }
}

impl From<serde_json::Error> for TypedError {
    fn from(err: serde_json::Error) -> Self {
        TypedError::Json(err)
    }
}

impl fmt::Display for TypedError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            TypedError::Ffi(err) => fmt::Display::fmt(err, f),
            TypedError::Json(err) => write!(f, "invalid Lite JSON response: {err}"),
        }
    }
}

impl std::error::Error for TypedError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            TypedError::Ffi(err) => Some(err),
            TypedError::Json(err) => Some(err),
        }
    }
}

/// A `Result` whose error is [`TypedError`], used by the typed convenience
/// methods (as opposed to [`crate::Result`], used by the raw `*_json`
/// methods).
pub type TypedResult<T> = std::result::Result<T, TypedError>;

/// Describes the physical Lite database file backing a handle.
#[derive(Debug, Clone, Deserialize)]
pub struct StorageStatus {
    pub format: String,
    pub engine: String,
    pub primary_layout: String,
    pub replay_layout: String,
    pub index_layout: String,
    #[serde(default)]
    pub index_namespace: Option<String>,
    #[serde(default)]
    pub format_version: Option<u32>,
    #[serde(default)]
    pub page_size: Option<u32>,
    #[serde(default)]
    pub active_checkpoint: Option<u8>,
    #[serde(default)]
    pub checkpoint_sequence: Option<u64>,
    #[serde(default)]
    pub page_count: Option<u64>,
}

/// Reports the configured inference execution mode for a Lite handle. See
/// `go/pkg/lite/status.go`'s `InferenceStatus` doc comment for the budget
/// field semantics (0 vs. "resolved" values).
#[derive(Debug, Clone, Deserialize)]
pub struct InferenceStatus {
    pub mode: String,
    pub available_modes: Vec<String>,
    pub configured: bool,
    pub remote_provider_configured: bool,
    pub local_runtime_configured: bool,
    pub local_runtime_available: bool,
    pub caller_supplied_artifacts: bool,
    pub no_inference_configured_ok: bool,
    pub host_budget_mb: u32,
    pub backend_budget_mb: u32,
    pub combined_budget_mb: u32,
    pub kv_budget_mb: u32,
    pub scratch_budget_mb: u32,
    pub process_memory_budget_mb: u32,
    pub process_memory_limit_bytes: u64,
    pub process_memory_limit_source: String,
}

/// Describes the Lite feature contract advertised by a handle.
#[derive(Debug, Clone, Deserialize)]
pub struct Capabilities {
    pub freestanding_build: bool,
    pub threading: String,
    pub hosted_profile: bool,
    pub manual_maintenance: bool,
    pub background_enrichment_runtime: bool,
    pub ttl_cleanup_runtime: bool,
    pub transaction_recovery_runtime: bool,
    pub local_template_rendering: bool,
    pub remote_template_rendering: bool,
    pub remote_template_host_callbacks: bool,
    pub inference_mode: String,
    pub supported_inference_modes: Vec<String>,
    pub available_inference_modes: Vec<String>,
    pub inference_required: bool,
    pub no_inference_configured_ok: bool,
    pub caller_supplied_artifacts: bool,
    pub caller_supplied_embeddings: bool,
    pub remote_inference_providers: bool,
    pub local_inference_runtime: bool,
    pub generated_enrichment_planning: bool,
    pub text_search: bool,
    pub dense_vector_search: bool,
    pub sparse_vector_search: bool,
    pub hybrid_search: bool,
    pub graph_search: bool,
    pub distributed_shard_ownership: bool,
    pub raft_replication: bool,
    pub cluster_placement: bool,
    pub cross_node_joins: bool,
    pub remote_shard_fanout: bool,
    pub distributed_transaction_coordination: bool,
    pub cluster_heartbeat_status_aggregation: bool,
    pub server_side_autoscaling: bool,
    pub kubernetes_operator: bool,
    pub object_storage_primary: bool,
}

/// Describes the stable Lite readiness fields for derived work. Nested
/// maintenance telemetry is kept as raw JSON (`serde_json::Value`) because
/// its shape is operational detail rather than a stable control-plane
/// contract.
#[derive(Debug, Clone, Deserialize)]
pub struct PendingWorkStatus {
    pub derived_target_sequence: u64,
    pub has_async_indexes: bool,
    pub portable_import_publication_in_progress: bool,
    pub portable_import_recovery_required: bool,
    pub portable_runtime_activation_pending: bool,
    #[serde(default)]
    pub enrichment: serde_json::Value,
    #[serde(default)]
    pub resolution: serde_json::Value,
    #[serde(default)]
    pub promotion: serde_json::Value,
    #[serde(default)]
    pub text_merge: serde_json::Value,
}

/// The typed form of [`Database::status_json`]. `stats` is kept as raw JSON
/// because its shape is broader than the stable Lite control fields
/// bindings need for feature branching.
#[derive(Debug, Clone, Deserialize)]
pub struct Status {
    pub storage: StorageStatus,
    #[serde(default)]
    pub stats: serde_json::Value,
    pub pending_work: PendingWorkStatus,
    pub inference: InferenceStatus,
    pub capabilities: Capabilities,
}

/// Reports how many generated enrichment references were recreated from
/// stored documents.
#[derive(Debug, Clone, Deserialize)]
pub struct ReplayGeneratedEnrichmentsResult {
    pub replayed: u64,
}

impl Database {
    /// Returns the typed Lite status document for the database.
    pub fn status(&self) -> TypedResult<Status> {
        let body = self.status_json()?;
        Ok(serde_json::from_slice(&body)?)
    }

    /// Returns the typed Lite pending-work readiness document.
    pub fn pending_work_stats(&self) -> TypedResult<PendingWorkStatus> {
        let body = self.pending_work_stats_json()?;
        Ok(serde_json::from_slice(&body)?)
    }

    /// Drains pending enrichment and index work and returns the typed
    /// post-drain readiness document.
    pub fn run_until_idle_status(&self) -> TypedResult<PendingWorkStatus> {
        let body = self.run_until_idle_json()?;
        Ok(serde_json::from_slice(&body)?)
    }

    /// Recreates generated enrichment work from stored documents and
    /// returns the replay count.
    pub fn replay_generated_enrichments(&self) -> TypedResult<ReplayGeneratedEnrichmentsResult> {
        let body = self.replay_generated_enrichments_json()?;
        Ok(serde_json::from_slice(&body)?)
    }

    /// Returns the typed Lite capability document for the database.
    pub fn capabilities(&self) -> TypedResult<Capabilities> {
        let body = self.capabilities_json()?;
        Ok(serde_json::from_slice(&body)?)
    }
}
